Optimize distinct

d-dorazio opened the pull request:
In this PR I've tried to speedup the distinct command by optimizing the simulated annealing process and the rearranging of colors.

Non scientific benchmarks show that in fact the proposed changes greatly improve performance. For example, now I'm able to run `pastel distinct 200 | cat` in <1 sec while without these changes it would have taken ~25secs on my machine. However, on small inputs the boost is not as big; in fact `pastel distinct 20 | cat` now runs in ~0.2secs vs ~0.4secs which is still better, but still a bit slow imho.

The simulated annealing was optimized mainly by avoiding to recalculate the distances from scratch every time , but by only recalculating the distances to the color that changed. This is the main contributing factor to the speedup.

After having optimized that, `cargo flamegraph` showed that a _bunch_ of time was spent inside the `rearrange_sequence` function. A first optimization was to simply switch to using `sort_by_cached_key` because calculating the key at each comparison is really too expensive. Then,  I went ahead and removed the sorting altogether because I think there was no need for it and introduced a cache for the minimum distances.

At this point I don't see any other low hanging fruits and I'm likely to stop here as I'm ok with the results.

Let me know what you think.

(also, your implementation of simulated annealing was really cool! simulated annealing is one of my favourite algorithms :smile:  )

sharkdp commented:
Thank you very much for working on this!

The speed improvements are impressive. Unfortunately, the behavior isn't quite the same and I can not reproduce the same results. This is a bit difficult to measure, as there is randomness involved, but you can try the following: Run `pastel distinct -v 6` and take a look at the optimized D_min score. For the old version, it should consistently be around 103.2. With your version, it also reaches 103.2 sometimes, but it also yields results which are much worse.

10 runs of the old version:
```
103.23
103.23
103.22
103.22
103.23
103.22
103.23
103.23
103.19
103.22
```

10 runs of the new version:
```
98.90
103.23
103.22
98.90
103.22
103.22
98.90
94.28
98.31
93.48
```

What I can also observe from running `pastel distinct` with the `-v`/`--verbose` option is that the `D_mean` score is excessively high in your version. It looks like a some `scalar::MAX` values accidentally made it into the computation for the mean.

I will add some inline comments and see if I can find any mistakes.

Thanks!

sharkdp commented on src/distinct.rs:
```diff
@@ -17,6 +18,11 @@ pub struct DistanceResult {
     /// Indices of the colors that were closest to each other
     pub closest_pair: (usize, usize),
+
+    /// The closest distance and color index for each color
```
minor: *"The closest distance and the index of the nearest neighbor"*?

sharkdp commented on src/distinct.rs:
```diff
@@ -227,13 +182,104 @@ pub fn rearrange_sequence(colors: &mut Vec<Color>, metric: DistanceMetric) {
+
+            if dist < self.closest_distances[i].0 {
+                self.closest_distances[i] = (dist, color);
+                self.closest_distances[color] = (dist, i);
```
I think there is a mistake here: If `dist < self.closest_distances[i].0`, it is okay to update `self.closest_distances[i]`. But that doesn't necessarily mean that `self.closest_distances[color]` should also be updated, right? `color` could be closer to another (third) color but still be the nearest neighbor for `i`.

I think there should be a separate `if` block for updating `self.closest_distances[color]`. This could also be the reason for `scalar::MAX` values appearing in the mean computation. Imagine that we shift `color` very far away from all other colors and then recompute the distances. Since it is not a nearest neighbor of any of the other colors, `self.closest_distances[color]` will never be updated in this loop and the value will still be `scalar::MAX`.

d-dorazio replied:
you're absolutely right, my bad I didn't catch the error :sweat_smile: 

I've added the separate check and it seems that the D_min is now as high as before. Occasionally I get some lower results like 98.3 but it seems to me that the current version of pastel also returns those values sometimes.

sharkdp replied:
> Occasionally I get some lower results like 98.3 but it seems to me that the current version of pastel also returns those values sometimes.

I'm sure that's possible. The really nice thing about your PR is that we can now probably adapt the simulated annealing parameters a bit because it is now so fast. I would like to increase the iteration counts (and simultaneously decrease the cooling rates). This should lead to much more stable results.

d-dorazio replied:
yeah, I didn't want to change the parameters because I think the plan is too actually expose those parameters as command line options, but I guess it's ok to tune the parameters now.

However, I think it would be interesting to have an "adaptive" metric in the simulated annealing process too that would potentially stop the process early if, say, the relative error is negligible or something.

sharkdp replied:
> However, I think it would be interesting to have an "adaptive" metric in the simulated annealing process too that would potentially stop the process early if, say, the relative error is negligible or something.

Absolutely

sharkdp commented on src/distinct.rs:
```diff
@@ -227,13 +182,104 @@ pub fn rearrange_sequence(colors: &mut Vec<Color>, metric: DistanceMetric) {
+        result
+    }
+
+    fn update_distances(&mut self, colors: &[(Color, Lab)], color: usize, changed: bool) {
```
minor: maybe rename `color` to `updated_color` or `modified_color`? Also, we could maybe rename `changed` to `perform_recursive_update`?

d-dorazio replied:
Initially I named something like that, but `update_distances` is really just a way to recalculate the distances and it can be called with colors that aren't actually changed (like in the case of `to_recalc`) so that's why I generally named the parameter `color`. Wrt `changed` instead I'm ok with calling it `perform_recursive_update` even though I think I still prefer `changed` because I think the caller of the functions is more likely to understand what `changed` means than `perform_recursive_update` because it would have to understand that `perform_recursive_update` needs to be done whenever the color has changed.

Anyway, if you prefer the new names I'm ok with changing them.

sharkdp replied:
No, that makes sense. Let's go with your naming :+1:

sharkdp commented:
> After having optimized that, `cargo flamegraph` showed that a _bunch_ of time was spent inside the `rearrange_sequence` function. A first optimization was to simply switch to using `sort_by_cached_key` because calculating the key at each comparison is really too expensive. Then, I went ahead and removed the sorting altogether because I think there was no need for it and introduced a cache for the minimum distances.

Using `sort_by_cached_key` sounds very reasonable, but I don't understand the second part. What do you mean by "there was no need for it"? I didn't really understand your new implementation when taking a first look, but I can try to take a closer look again.

d-dorazio commented:
> > After having optimized that, `cargo flamegraph` showed that a _bunch_ of time was spent inside the `rearrange_sequence` function. A first optimization was to simply switch to using `sort_by_cached_key` because calculating the key at each comparison is really too expensive. Then, I went ahead and removed the sorting altogether because I think there was no need for it and introduced a cache for the minimum distances.
> 
> Using `sort_by_cached_key` sounds very reasonable, but I don't understand the second part. What do you mean by "there was no need for it"? I didn't really understand your new implementation when taking a first look, but I can try to take a closer look again.

I'm sorry it wasn't clear, I could have explained better.

My understanding of the algorithm is that it has to rearrange the sequence of a colors in such a way that the distance between a color and its predecessors is maximized.

My idea was to do something similar to insertion sort where you search for the element to insert at position i by finding the minimum of the elements from i+1 to the end of the input. I said there is no need to sort the "right" part of the sequence because I think you really need to just find the minimum which is O(n) vs O(n log(n)).

Moreover at each step we can be clever and avoid recomputing the minimal distance from each element in the "right" part to each element in the "left" one because we know which element we inserted in the previous run of the loop and so we can update the previous differences with only the new value.

sharkdp commented on src/distinct.rs:
```diff
@@ -227,13 +182,107 @@ pub fn rearrange_sequence(colors: &mut Vec<Color>, metric: DistanceMetric) {
         DistanceMetric::CIEDE2000 => c1.distance_delta_e_ciede2000(c2),
     };
 
+    // vector where the i-th element contains the minimum distance to the colors from 0 to i-1.
```
Should this be "contains the minimum distance to the colors from i+1 to n"?

d-dorazio replied:
I think it's correct the way it is. The element at position i is the minimum distance to the current left part of the colors which is 0..i . Note that I've inverted a bit the previous logic where it was using `-` and `max` to take the minimum of the distances to the color and then sorting in descending order. Hopefully now this logic should be a bit more clear and matches the description of the algorithm.

sharkdp commented:
Thank you for the clarification. I somehow missed that the outer loop was still there and got confused as to how that still works :smile:. Yes, that makes sense! Thank you for the clarification.

If I understand it correctly, your changes brings the complexity from *O(N² log N)* down to *O(N²)*.

> Moreover at each step we can be clever and avoid recomputing the minimal distance from each element in the "right" part to each element in the "left" one because we know which element we inserted in the previous run of the loop and so we can update the previous differences with only the new value.

This part I still don't quite understand. We did that before as well, right? I only called `sort_unstable_by_key` on the `right` part. In your implementation, wouldn't it be the same if we just used `min_by_key` on the right part?

d-dorazio commented:
> If I understand it correctly, your changes brings the complexity from O(N² log N) down to O(N²).

This is correct if we ignore how the calculation of the keys was done. Previously for each element in "right" it was iterating over each element in "left" to find the maximum which I believe is O(N²) (but I never understood big O :) ) so the final complexity of the previous algorithms (I think) was O(__N²__ N² log N) but now it's O(N²) because this calculation is cached and then updated at each step of the outer loop.

sharkdp commented:
Ah, right. That explains why it was so slow for larger N :smile:

sharkdp commented:
Thank you!

d-dorazio commented:
Thank you for merging so quickly! :rocket:  :tada:

sharkdp commented:
Released in [v0.6.0](https://github.com/sharkdp/pastel/releases).
